### ISSUE TITLE
Adding namespace-less node to namespaced parent attaches the parent namespace to the child

### ISSUE THREAD

**Issue by mbklein** [sparklemotion/nokogiri#425]
Start with

```
doc = Nokogiri::XML(%{
<my:document xmlns:my="foo:bar">
  <my:first-child>Some Text!</my:first-child>
</my:document>
})
```

Then

```
doc.root.add_child(%{<second-child/>})
```

or

```
doc.root.add_child(%{<third-child xmlns=""/>})
```

or even

```
new_child = Nokogiri::XML(%{<fourth-child/>}).root
doc.root.add_child(new_child)
```

Do it all, and you end up with

```
<?xml version="1.0"?>
<my:document xmlns:my="foo:bar">
  <my:first-child>Some Text!</my:first-child>
  <my:second-child/>
  <my:third-child xmlns=""/>
  <my:fourth-child/>
</my:document>
```

None of which is probably what you meant.

The only way I've been able to get the new element (and its children) to end up with no namespace is to traverse it after adding and explicitly set `node.namespace = nil` on everything.

**Comment by jav20a** [comment]
Hi. Is this resolved?

**Comment by pgericson** [comment]
This does not seem to be resolved, I'm on version `1.8.0`.

**Comment by mmustala** [comment]
What a surprising and annoying feature. I would have expected `node.replace(xml_string)` to just change the content of the node, but no, it messes up the namespaces. I guess I need to write a regex to get the node replaced properly.

**Comment by jdongelmans** [comment]
FYI: My downstream gem broke because of a now missing namespace and I stumbled on this old issue because with the upgrade to version `>= 1.12.0` (and thus when `libxml` version `>= 2.9.12`, ref: https://github.com/sparklemotion/nokogiri/commit/889ee2a9cb1e190bfa664cbf3552585f4d0a09a7) this has been fixed.

**Comment by flavorjones** [comment]
@jdongelmans Thanks for the note. I'd like to clarify, though: 889ee2a noted a change in the HTML parser, but this issue is about XML parser behavior.

That said, this behavior has also changed in the XML parser (for some of the cases described), and I had a mental note to circle back to this issue to verify that and note it here. I *think* the behavior changed in #2230 (specifically 1f483f0) but will verify that.

**Comment by flavorjones** [comment]
OK, after a bit of git-bisecting ...

The "third child" example already worked as expected in v1.6.0 which is as far back as I can easily go on my dev machine.

But the "second child" and "fourth child" cases were (as suspected) fixed by 1f483f0 which first appeared in v1.12.0 and was part of #2230 which fixed #1712 and #2228. See also #2246 which updated some tests to reflect this behavioral change.

So: I'm closing this one out as "fixed in v1.12.0". Sorry for the delay and the confusion caused by leaving this open.

**Comment by flavorjones** [comment]
Please see #2320 for more discussion about this change.
